### PR TITLE
Move QueryStats tracking into the QueryEngine and out of the TabletPlan 

### PR DESCRIPTION
**Pull request by zmagg**
### Description
The QueryStats metrics (including `QueryCounts`) are incorrect and unreliable for our keyspaces with large amounts of bulk inserts. This is because the TabletPlan cache evicts frequently and the query stats are driven out of the TabletPlan cache. So, on every given metric scrape, the plan in question is not reliably in the cache (and therefore its stats are not there) to be scraped, leading to false gaps in metrics that Prometheus struggles to compute `rate()`s off of.

This results in graphs that look like this (out of Grafana/Prometheus), when we did not actually have those spikes in query rates on that tablet:
screen shot 2018-07-18 at 5 42 46 pm

One solution to this problem is proposed in https://github.com/vitessio/vitess/issues/3667 . If we normalized bulk insert queries, there would be fewer unique queries in the cache and they would rarely get evicted. 

This PR proposes that we do something simpler for now, which is just to stash the QueryStats in the QueryEngine and not in the TabletPlan, therefore not evicting the stats when we evict tablet plans.

### CODE REVIEW

**Comment by sougou** [comment]
I think this is potentially dangerous because there's no upper limit on the number of unique queries. So it can cause vttablet to OOM.

One possible hack-around till we get to #3667 is as follows: After building the plan, if the of number inserted rows >X, don't add it to the query stats at all.
The downside is that we won't know the stats for bulk inserts.

We could also create a fake `BULK_INSERT` entry. But it may not be worth it.

**Comment by zmagg** [comment]
@sougou That makes sense. Yeah. For our particular case it seemed safe-ish because the cardinality of query plan string + table name is very low on all our keyspaces but I definitely see that it doesn't make sense generically.

Hm. I also think that #3667 is just a better solution (should also have some perf wins as well). 

What do you mean by a fake `BULK_INSERT`? A fake general `BULK_INSERT` type in the query plan cache?

**Comment by zmagg** [comment]
@sougou Curious what you think about this and if this sounds right to you / I'm missing anything:

@demmer and I have been talking about this again recently, as it continues to cause pain for us. 

He pointed out that there _is_ a baked in upper limit on the number of queries in QueryStats (keyed off of PlanType+TableName) as only queries that are against tables in the database schema will ever be registered in QueryStats. So, a user couldn't erroneously issue a large number of junk queries on junk tables and OOM the vttablet via the QueryStats data structure.

I both tested this with a junk query on an undefined table name and traced the [code](https://github.com/tinyspeck/vitess/blob/master/go/vt/vttablet/tabletserver/query_engine.go#L443) and this seems right to me. 

He also suggested some perf improvements to the QueryStats data structure, which I've committed to this PR.

Thoughts?

**Comment by zmagg** [comment]
!! The endtoend test failures taught me about the existence of the `debug/query_stats` page which I had never seen before. 

My previous comment is still true, but might be moot--I've now added the sql query string to the key of the  `QueryStats` datastructure in order to power the `debug/query_stats` page. 

The other idea that @demmer and I were talking about was just applying a LRU to the QueryStats datastructure. As there are significantly fewer objects in QueryStats than there are in TabletPlans, a reasonable LRU should mitigate OOM-risks and still keep our metrics from over-evicting. What do you think?

**Comment by zmagg** [comment]
(And I’ve just noticed that some of the tests are still failing, I’ll take a look and fix that in the morning.)

File: go/vt/vttablet/tabletserver/query_executor.go [hunk 1]
```diff
@@ -80,10 +85,10 @@ func (qre *QueryExecutor) Execute() (reply *sqltypes.Result, err error) {
 		tabletenv.RecordUserQuery(qre.ctx, qre.plan.TableName(), "Execute", int64(duration))
 
 		if reply == nil {
-			qre.plan.AddStats(1, duration, qre.logStats.MysqlResponseTime, 0, 1)
+			qre.addStats(planName, 1, duration, qre.logStats.MysqlResponseTime, 0, 1)
 			return
 		}
-		qre.plan.AddStats(1, duration, qre.logStats.MysqlResponseTime, int64(reply.RowsAffected), 0)
+		qre.addStats(planName, 1, duration, qre.logStats.MysqlResponseTime, int64(reply.RowsAffected), 0)
```
Comment (demmer) [inline go/vt/vttablet/tabletserver/query_executor.go]: This is somewhat stylistic, but IMO we should move the `qre.logStats` and `tabletenv.ResultStats` updating inside the new `qre.addStats` function to keep all the query metrics and log recording together.

Comment (demmer) [reply go/vt/vttablet/tabletserver/query_executor.go]: (Or don't have the function at all and just keep all the updating in-line... it's having a helper function that does some but not all of the stats work which I feel like could be cleaned up)

File: go/vt/vttablet/tabletserver/query_executor.go [hunk 2]
```diff
@@ -69,6 +69,11 @@ var sequenceFields = []*querypb.Field{
 	},
 }
 
+func (qre *QueryExecutor) addStats(planName string, queryCount int64, duration, mysqlTime time.Duration, rowCount, errorCount int64) {
+	qre.tsv.qe.AddStats(planName, qre.plan.TableName().String(), queryCount, duration, mysqlTime, rowCount, errorCount)
```
Comment (demmer) [inline go/vt/vttablet/tabletserver/query_executor.go]: We need some handling for the case where the table name is unknown...

For consistency it should use "Join" as the "table name" in that case.

Comment (zmagg) [reply go/vt/vttablet/tabletserver/query_executor.go]: Hmm. Where there's an unknown table identifier, it looks like qre.plan.TableName().String() is just the empty string which I think will be fine for our metrics. Would you prefer "Join" over that?

**Comment by sougou** [comment]
Was the OOM concern addressed?

**Comment by demmer** [comment]
Echoing a conversation from Slack, there isn't really an OOM concern here since the key is TableName + PlanType and we already have data structures in the tablet that are O(Number of Tables).

So I think this is fine.

**Pull request merged**